projects/bootloader/scripts/flash_application_code.py

```diff
@@ -0,0 +1,153 @@
+"""This script handles the client side of the flash application code protocol."""
+
+# pylint: disable=C0411
+# pylint: disable=C0413
+# pylint: disable=W0212
+# autopep8: off
+
+from can_datagram import Datagram, DatagramListener, DatagramSender
+import can
+import zlib as zl
+import time
+import os
+import sys
+# Import protobuf
+currentdir = os.path.dirname(os.path.realpath(__file__))
+parentdir = os.path.dirname(currentdir)
+sys.path.append(parentdir)
+from protogen import flash_application_code_pb2 as pb2
+
+OUTGOING_PROTOBUF_ID = 8
+OUTGOING_DATA_ID = 9
+
+
+def flash_application_code(**kwargs):
+    """Full Flash Application Code Function"""
+
+    flash = FlashApplication(**kwargs)
+
+    flash.flash_protobuf()
+    chunked_application_code = flash.chunks(flash.app_data, 2048)
+    for chunk in chunked_application_code:
+        # Reset recieved keys in listener
+        flash.listener.datagram_messages = {}
+        flash.recv_boards = []
+        flash.flash_application_chunk(chunk)
+
+
+class FlashApplication:
+    """Flash Application Code Protocol"""
+
+    # pylint: disable=too-many-instance-attributes
+    def __init__(self, **kwargs):
+        self._check_kwargs(**kwargs)
+
+        self.channel = kwargs["channel"]
+        self.app_data = kwargs["application_code_data"]
+        self.board_ids = kwargs["board_ids"]
+        self.name = kwargs["name"]
+        self.git_version = kwargs["git_version"]
+
+        self.sender = DatagramSender(channel=self.channel, receive_own_messages=False)
+        self.listener = DatagramListener(self.listener_callback)
+        self.notifier = can.Notifier(self.sender.bus, [self.listener])
+
+        self.recv_boards = []
+        self.status_code = 0
+
+    def listener_callback(self, datagram: Datagram, board_id):
+        """Check whether the datagram has an OK status code"""
+        # Ignores non-zero codes from the client
+        if datagram._data != bytearray(0) and board_id != 0:
```

Comment:
`bytearray(0)` creates an empty bytearray (length 0) - you probably want `bytearray([0])` to check for a length-1 bytearray containing a 0.